prow/plank/reconciler.go

```diff
@@ -781,29 +792,39 @@ func (r *reconciler) canExecuteConcurrently(ctx context.Context, pj *prowv1.Prow
 	}
 	r.log.Infof("got %d not completed with same name", len(pjs.Items))
 
-	var pendingOrOlderMatchingPJs int
-	for _, foundPJ := range pjs.Items {
-		// Ignore self here.
-		if foundPJ.UID == pj.UID {
-			continue
-		}
-		if foundPJ.Status.State == prowv1.PendingState {
-			pendingOrOlderMatchingPJs++
-			continue
-		}
+	pendingOrOlderMatchingPJs := countPendingOrOlderMatchingPJs(*pj, pjs.Items)
+	if pendingOrOlderMatchingPJs >= pj.Spec.MaxConcurrency {
+		r.log.WithFields(pjutil.ProwJobFields(pj)).
+			Debugf("Not starting another instance of %s, have %d instances that are pending or older, %d is the limit",
+				pj.Spec.Job, pendingOrOlderMatchingPJs, pj.Spec.MaxConcurrency)
+		return false, nil
+	}
 
-		// At this point we know that foundPJ is in Triggered state, if its older than our prowJobs it gets
-		// priorized to make sure we execute jobs in creation order.
-		if foundPJ.CreationTimestamp.Before(&pj.CreationTimestamp) {
-			pendingOrOlderMatchingPJs++
-		}
+	return true, nil
+}
 
+func (r *reconciler) canExecuteConcurrentlyPerQueue(ctx context.Context, pj *prowv1.ProwJob) (bool, error) {
+	queueName := pj.Spec.JobQueueName
+	if queueName == "" {
+		return true, nil
 	}
 
-	if pendingOrOlderMatchingPJs >= pj.Spec.MaxConcurrency {
+	queueConcurrency, queueDefined := r.config().Plank.JobQueueConcurrencies[queueName]
+	if !queueDefined || queueConcurrency == 0 {
```

Comment:
We shouldn't allow/expect the `!queueDefined` case. Please add job config validation logic to the `config` package to ensure that all job queue names specified on jobs are associated with an entry in the `JobQueueConcurrencies` map. Including the validation logic there will catch such misconfigurations at presubmit time rather than runtime.

Reply:
If I understand correctly, the validation happens when a `JobConfig` is converted to a `ProwJob`. When a user does that locally, they can have the names already added to the prow configuration and the validation would pass. But that doesn't change the fact that if the server has outdated plank configuration, it would fail to find queue concurrency for given `queueName`. So we should anticipate for such a scenario either way, right?

Reply:
Another thing is that adding this validation logic to https://github.com/kubernetes/test-infra/blob/55d9a8ef93e1eccd103b77a00f96d7a84cf0cbbd/prow/config/config.go#L1905-L1943 might be a bit... "dirty". We would need to pass the `c.ProwConfig.Plank.JobQueueConcurrencies` down to `validatePresubmits`, `validatePostsubmits` and `validatePeriodics`.

Reply:
> If I understand correctly, the validation happens when a `JobConfig` is converted to a `ProwJob`

No, config validation happens when the config is loaded. We load the config as part of `checkconfig`'s validation at presubmit time to prevent submission of invalid config.

> When a user does that locally, they can have the names already added to the prow configuration and the validation would pass. But that doesn't change the fact that if the server has outdated plank configuration, it would fail to find queue concurrency for given queueName.

Job configuration and the core config are both updated immediately after the config change PR is merged. Plank should never have configuration that is outdated with respect to the job configuration by more than a few seconds and that can only happen after a job is added or queue name is changed.

> So we should anticipate for such a scenario either way, right?

We do need to gracefully handle this case. We don't want plank to explode in the rare event of a job being triggered and transitioning to pending immediately after the job is added and after the job config has been updated, but before the core config has been updated. We should log a debug level message and probably abort the job.

> Another thing is that adding this validation logic to [...] might be a bit... "dirty".

Yes, we'll have to do something like you suggested since the `JobQueueConcurrencies` is not available in `validateJobBase` currently, but the dirtiness is definitely worth the UX improvement of having the validation to catch this at presubmit time. Either of the approaches you mentioned seems appropriate for now, we can refactor later if we find more fields need to be passed down.

Reply:
Done. Made `validateJobBase` a method of `Config`. Thank you for explaining the existing logic to me.